mne_sandbox/connectivity/cfc.py

```diff
@@ -0,0 +1,506 @@
+    if npad == 'auto':
+        next_pow_2 = int(np.ceil(np.log2(n_times)))
+        npad = 2**next_pow_2 - n_times
+    elif not isinstance(npad, int):
+        raise ValueError('npad must be "auto" or a positive integer')
+
+    data_ph = np.hstack([data[ix_ph, :], np.zeros([len(ix_ph), npad])])
+    data_ph = filterfn(data_ph, sfreq, *f_ph, **kws_filt)
+    data_ph = np.angle(hilbert(data_ph))[..., :n_times]
+    ix_map_ph = {ix: i for i, ix in enumerate(ix_ph)}
+
+    data_am = np.hstack([data[ix_am], np.zeros([len(ix_am), npad])])
+    data_am = filterfn(data_am, sfreq, *f_am, **kws_filt)
+    data_am = np.abs(hilbert(data_am))
+    if hi_phase is True:
+        data_am = filterfn(data_am, sfreq, *f_ph, **kws_filt)
+        data_am = np.angle(hilbert(data_am))
+    data_am = data_am[..., :n_times]
+    ix_map_am = {ix: i for i, ix in enumerate(ix_am)}
+
+    if scale_amp_func is not None:
+        data_am = scale_amp_func(data_am, axis=-1)
+    return data_ph, data_am, ix_map_ph, ix_map_am
+
+
+def _filter_ph_am(xph, xam, f_ph, f_am, sfreq, filterfn=None, kws_filt=None):
+    """Aux function for phase/amplitude filtering for one pair of channels"""
+    from ..externals.pacpy.pac import _range_sanity
+    from scipy.signal import hilbert
+    filterfn = band_pass_filter if filterfn is None else filterfn
+    kws_filt = {} if kws_filt is None else kws_filt
+
+    # Filter the two signals + hilbert/phase
+    _range_sanity(f_ph, f_am)
+    xph = filterfn(xph, sfreq, *f_ph)
+    xam = filterfn(xam, sfreq, *f_am)
+
+    xph = np.angle(hilbert(xph))
+    xam = np.abs(hilbert(xam))
+    return xph, xam
+
+
+def _array_raw_to_epochs(x, sfreq, ev, tmin, tmax):
+    """Aux function to create epochs from a 2D array"""
+    if ev.ndim != 1:
+        raise ValueError('ev must be 1D')
+    if ev.dtype != int:
+        raise ValueError('ev must be of dtype int')
+
+    # Check that events won't be cut off
+    n_times = x.shape[-1]
+    min_ix = 0 - sfreq * tmin
+    max_ix = n_times - sfreq * tmax
+    msk_keep = np.logical_and(ev > min_ix, ev < max_ix)
+
+    if not all(msk_keep):
+        print('Some events will be cut off!')
+        ev = ev[msk_keep]
+
+    # Pull events from the raw data
+    epochs = []
+    for ix in ev:
+        ix_min, ix_max = [ix + int(i_tlim * sfreq)
+                          for i_tlim in [tmin, tmax]]
+        epochs.append(x[np.newaxis, :, ix_min:ix_max])
+    epochs = np.concatenate(epochs, axis=0)
+    times = np.arange(epochs.shape[-1]) / float(sfreq) + tmin
+    return epochs, times, msk_keep
+
+
+# For the viz functions
+def _extract_phase_and_amp(data_phase, data_amp, sfreq, freqs_phase,
+                           freqs_amp, scale=True):
+    """Extract the phase and amplitude of two signals for PAC viz.
+    data should be shape (n_epochs, n_times)"""
+    from sklearn.preprocessing import scale
+
+    # Morlet transform to get complex representation
+    band_ph = cwt_morlet(data_phase, sfreq, freqs_phase)
```

Comment:
Since this function is just for visualization. We could use Hilbert here too, but it'd be slower. These functions should probably be named differently to make that clear, but I didn't change it yet since these are all under-the-hood functions